233 Update to use latest metrics API and add compute_metric example

Nic-Ma opened the pull request:
Fixes #233 .

### Description
This PR updated the 3D segmentation examples to show how to use latest metrics APIs (@yiheng-wang-nv will help update all the other examples and tutorials later.)
And also added example to compute metrics based on saved predictions and labels with multi-processing, and generate the metrics report.

### Status
**Ready**

### Checks
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Notebook runs automatically `./runner [-p <regex_pattern>]`

Nic-Ma commented:
Hi @wyli ,

I updated the 3D segmentation examples to use the latest metrics APIs. And I didn't use notebook to show multi-processing computation of metrics because we need to launch python program with `dist.launch` and no chart need to show.
Could you please help review it?

Thanks in advance.

wyli commented on 3d_segmentation/torch/compute_metric.py:
```diff
@@ -0,0 +1,131 @@
+    transforms = Compose(
+        [
+            LoadImaged(keys=["pred", "label"]),
+            EnsureChannelFirstd(keys=["pred", "label"]),
```
looks good, could you make synthetic images in this demo to make it runnable? would be great to have some postprocessing transforms here such as `KeepLargestConnectedComponentD` it is often used during evaluation

Nic-Ma replied:
Hi @wyli ,

Thanks for your review.
This demo is runnable, it uses the saved `preds` and `labels` from the `unet_evaluation_dict.py` program, do you mean to copy an inference logic here?

Thanks.

wyli replied:
no, I mean having a simple and self-contained example for metrics computation

Nic-Ma replied:
So you mean to do below steps?
1. generate synthetic `images` and `labels`.
2. run evaluation with `images` and `labels`.(with `KeepLargestConnectedComponentD` as post transform)
3. save the `predictions` and `labels`.
4. compute metrics from saved `predictions` and `labels`.
5. generate metrics report.

Thanks.

wyli replied:
1. generate predictions and labels (with KeepLargest on predictions)
2. compute metrics
3. generate a report
this could be a new script in the modules/ folder

Nic-Ma replied:
OK, I see, let me investigate how to generate `preds` and `labels` directly, I thought our synthetic API can only generate `(image, label)` pair.

Thanks.

Nic-Ma replied:
updated in the latest commit.
Thanks.

Nic-Ma commented:
Hi @wyli ,

I have changed to use synthetic `preds` and `labels`, added `KeepLargestConnectedComponentd` according to your comments. Could you please help review it again?
For example, some test result of ` mean_dice_raw.csv`:
```
filename        class0  mean
./output/pred0.nii.gz   0.8550655       0.8550655
./output/pred10.nii.gz  0.74448234      0.74448234
./output/pred12.nii.gz  0.8262536       0.8262536
./output/pred14.nii.gz  0.96275276      0.96275276
./output/pred2.nii.gz   0.8499785       0.8499785
./output/pred4.nii.gz   0.99761885      0.99761885
./output/pred6.nii.gz   0.97747666      0.97747666
./output/pred8.nii.gz   0.9370028       0.9370028
./output/pred1.nii.gz   0.88721704      0.88721704
./output/pred11.nii.gz  0.94229424      0.94229424
./output/pred13.nii.gz  0.985372        0.985372
./output/pred15.nii.gz  0.92427427      0.92427427
./output/pred3.nii.gz   0.9563079       0.9563079
./output/pred5.nii.gz   0.9341649       0.9341649
./output/pred7.nii.gz   0.8643773       0.8643773
./output/pred9.nii.gz   0.56778765      0.56778765
```

Thanks.

wyli commented:
thanks, for the metrics module, is it possible to have one command to run single node multiprocess metric computation? that'll be immediately useful. also, I don't quite understand why `nccl` is needed in the demo

Nic-Ma commented:
Hi @wyli ,

This example code can support multiple-nodes or single-node, let me update the doc-string to put the example command to run `single node multiprocess` in the doc-string.
Let me change to GLOO to compute on CPU directly.

Thanks.

Nic-Ma commented:
Hi @wyli ,

Thanks for your suggestion, I updated the example based on your comments.
Could you please help review it again?

Thanks.

wyli commented on modules/compute_metric.py:
```diff
@@ -0,0 +1,143 @@
+
+    if args.local_rank == 0:
+        print("mean dice: ", result)
+        # generate metric report
```
`# generate metrics at output/mean_dice_raw.csv, output/mean_dice_summary.csv, output/metrics.csv`

Nic-Ma replied:
updated in the latest commit.

Thanks.

wyli commented on modules/compute_metric.py:
```diff
@@ -0,0 +1,143 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-d", "--dir", default="./output", type=str, help="root directory of labels and predictions.")
+    # must parse the command-line argument: ``--local_rank=LOCAL_PROCESS_RANK``, which will be provided by DDP
+    parser.add_argument("--local_rank", type=int)
```
could default to 0? could you add `help=...`

Nic-Ma replied:
This arg is necessary for `torch.distributed.launch`, users don't need to use it in the command, PyTorch will use this arg when calling this program:
https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py#L70

Thanks.

wyli reviewed (commented):
some comments inline

Nic-Ma commented on modules/compute_metric.py:
```diff
@@ -0,0 +1,143 @@
+It can even run on multi-nodes.
+Main steps to set up the distributed data parallel:
+
+- Execute `torch.distributed.launch` to create processes on every node for every process.
```
Thanks for let me know this, in case to support PyTorch 1.5 - 1.8.1, I think we can still use `launch` here.
Will use `run` instead in future.

Thanks.

Nic-Ma commented on modules/compute_metric.py:
```diff
@@ -0,0 +1,143 @@
+# limitations under the License.
+
+"""
+This example shows how to compute metric in multi-processing based on PyTorch native `DistributedDataParallel` module.
```
updated in the latest commit.
Thanks.

Nic-Ma commented on modules/compute_metric.py:
```diff
@@ -0,0 +1,143 @@
+Note:
+    `torch.distributed.launch` will launch `nnodes * nproc_per_node = world_size` processes in total.
+    Example script to execute this program on a single node with 2 processes:
+    python -m torch.distributed.launch --nproc_per_node=2 --nnodes=1 --node_rank=0
```
Yes, we can just use the default values for this example, updated in the latest commit.

Thanks.
